core/src/test/scala/unit/kafka/log/LogManagerTest.scala

```diff
@@ -563,4 +565,39 @@ class LogManagerTest {
     logManager.topicConfigUpdated("test-topic")
     assertTrue(logManager.partitionsInitializing.isEmpty)
   }
+
+  @Test
+  def testMetricsExistWhenLogIsRecreatedBeforeDeletion(): Unit = {
+    val topicName = "metric-test"
+    def logMetrics = KafkaYammerMetrics.defaultRegistry.allMetrics.keySet.asScala.
+      filter(_.getType == "Log").filter(_.getScope.contains(topicName))
+
+    val tp = new TopicPartition(topicName, 0)
+    logManager.getOrCreateLog(tp, () => logConfig)
+
+    val metricTag = s"topic=${tp.topic},partition=${tp.partition}"
+
+    assertEquals(LogMetricNames.allMetricNames.size, logMetrics.size)
+    logMetrics.foreach { metric =>
+      assertTrue(metric.getMBeanName.contains(metricTag))
+    }
+
+    logManager.asyncDelete(tp)
+
+    assertTrue(logMetrics.isEmpty)
+
+    logManager.getOrCreateLog(tp, () => logConfig)
+
+    assertEquals(LogMetricNames.allMetricNames.size, logMetrics.size)
+    logMetrics.foreach { metric =>
+      assertTrue(metric.getMBeanName.contains(metricTag))
+    }
+
+    time.sleep(logConfig.fileDeleteDelayMs + 1)
```

Comment:
Why is this sleep needed?

Reply:
It causes enough "time" to elapse that the actual file deletion is triggered. Having it here tests that we still have metrics after the deletion occurs.

Reply:
Is it deterministic? I realize its mocktime, but threads can be scheduled in any order. Or using `MockTime` means all schedulers will get triggered before next line in test will get run?

Reply:
Yes, `MockTime` contains a `MockScheduler`, which explicitly checks for any scheduled tasks that need to be run whenever we sleep. And we pass that `MockScheduler` into the `Log` and `LogManager` constructors in the tests. See https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/utils/MockScheduler.scala#L59.